In [354]:
import pandas as pd

In [355]:
dif = pd.read_csv("//Difford's_Guide.csv")

In [356]:
dif.head()

Name  \
0                                April Fool   
1                        Brooklyn (perfect)   
2                     Californian Margarita   
3  Calvados Fruit Cup (Difford's Cup No.13)   
4                  Clover Club (House-Made)   

                                         Ingredients  \
0                           ['3 shot Chilled water']   
1  ['1 1⁄2 shot Straight rye whiskey (100 proof /...   
2  ['1 1⁄2 shot Patrón Silver blanco tequila', '3...   
3  ['5⁄6 shot Avallen Calvados', '5⁄6 shot Orange...   
4  ['1 1⁄2 shot Rutte Dry Gin', '1 1⁄2 spoon Rasp...   

                                              Making  \
0       STIR with ice and strain into chilled glass.   
1  STIR all ingredients with ice and strain into ...   
2  SHAKE all ingredients with ice and strain into...   
3  POUR all ingredients into glass. Half fill gla...   
4  STIR jam with gin in base of shaker until diss...   

                                                Link                Glass  \
0  https://www.diffordsguide.com/cocktails/recipe...        Martini glass   
1  https://www.diffordsguide.com/cocktails/recipe...          Coupe glass   
2  https://www.diffordsguide.com/cocktails/recipe...  Old-fashioned glass   
3  https://www.diffordsguide.com/cocktails/recipe...        Collins glass   
4  https://www.diffordsguide.com/cocktails/recipe...          Coupe glass   

                                             Garnish          Type  \
0               Lemon zest twist (discarded) & olive  Nonalcoholic   
1                          Luxardo maraschino cherry           NaN   
2                   Salt rim (optional) & lime wedge           NaN   
3  Lemon, orange & strawberry slices, mint sprig ...           NaN   
4  Lemon zest twist (expressed and discarded) & s...           NaN   

                                              Flavor  
0                                                NaN  
1  ['Gentle 0 - Boozy 10: 8', 'Sweet 0 - Dry/Sour...  
2  ['Gentle 0 - Boozy 10: 7', 'Sweet 0 - Dry/Sour...  
3  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...  
4  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...

In [357]:
dif.describe()
# 이름이 다른데 재료가 같은 칵테일 발견. 나중에 자세히 확인할 것.

Name                                        Ingredients  \
count         5339                                               5339   
unique        5339                                               5333   
top     April Fool  ['2 1⁄2 shot Rutte Dry Gin', '1⁄2 shot Martini...   
freq             1                                                  3   

                                                   Making  \
count                                                5337   
unique                                               1774   
top     SHAKE all ingredients with ice and fine strain...   
freq                                                 1342   

                                                     Link        Glass  \
count                                                5339         5339   
unique                                               5339           47   
top     https://www.diffordsguide.com/cocktails/recipe...  Coupe glass   
freq                                                    1         1843   

                  Garnish          Type  \
count                5339            34   
unique               1624             1   
top     Orange zest twist  Nonalcoholic   
freq                  644            34   

                                                   Flavor  
count                                                3064  
unique                                                 79  
top     ['Gentle 0 - Boozy 10: 7', 'Sweet 0 - Dry/Sour...  
freq                                                  449

In [358]:
difcopy = dif.copy()

### 중복 확인 및 제거

전체 중복 확인

In [359]:
difcopy_ent_dup = difcopy.duplicated(subset=None, keep=False)
len(difcopy_ent_dup[difcopy_ent_dup==True])

0

이름+재료만 중복 확인

In [360]:
difcopy_2_dup = difcopy.duplicated(["Name", "Ingredients"], keep=False)
len(difcopy_2_dup[difcopy_2_dup==True])

0

### Ingredients 형식 맞추기

In [361]:
print(difcopy["Ingredients"][1])

['1 1⁄2 shot Straight rye whiskey (100 proof / 50% alc./vol.)', '1⁄2 shot Martini Extra Dry vermouth', '1⁄2 shot Martini Rosso sweet vermouth', '1⁄3 shot Amer Picon', '1⁄3 shot Luxardo Maraschino liqueur', '1⁄3 shot Chilled water (omit if using wet ice)']


양 끝의 '['와 ']' 없애기, ' 없애기, shot을 oz로 바꾸기

In [362]:
for i in range(len(difcopy)):
    ingredients = difcopy["Ingredients"][i][1:-1]
    difcopy["Ingredients"][i] = ingredients

In [363]:
# 확인
print(difcopy["Ingredients"][1])

'1 1⁄2 shot Straight rye whiskey (100 proof / 50% alc./vol.)', '1⁄2 shot Martini Extra Dry vermouth', '1⁄2 shot Martini Rosso sweet vermouth', '1⁄3 shot Amer Picon', '1⁄3 shot Luxardo Maraschino liqueur', '1⁄3 shot Chilled water (omit if using wet ice)'


In [364]:
for i in range(len(difcopy)):
    ingredient_list = difcopy["Ingredients"][i].split(", ")
    for j in range(len(ingredient_list)):
        if 'shot' in ingredient_list[j].lower():
            if ingredient_list[j].lower().split(" ")[1]!='shot' and ingredient_list[j].lower().split(" ")[2]!='shot':
                print(i)

In [365]:
for i in range(len(difcopy)):
    ingredient_list = difcopy["Ingredients"][i].split(", ")
    for j in range(len(ingredient_list)):
        if 'shot' in ingredient_list[j].lower():
            ingredient_list[j] = ingredient_list[j].replace("shot", "oz")
        if ingredient_list[j][0]=="'":
            ingredient_list[j] = ingredient_list[j][1:]
        if ingredient_list[j][-1]=="'":
            ingredient_list[j] = ingredient_list[j][:-1]
    difcopy["Ingredients"][i] = ", ".join(ingredient_list)

In [366]:
# 확인
print(difcopy["Ingredients"][1])

1 1⁄2 oz Straight rye whiskey (100 proof / 50% alc./vol.), 1⁄2 oz Martini Extra Dry vermouth, 1⁄2 oz Martini Rosso sweet vermouth, 1⁄3 oz Amer Picon, 1⁄3 oz Luxardo Maraschino liqueur, 1⁄3 oz Chilled water (omit if using wet ice)


### Making 형식 맞추기

In [367]:
difcopy["Making"].isnull().sum()
# Making이 공백인 데이터 발견.

2

In [368]:
difcopy[difcopy["Making"].isnull()]
# 원래 사이트를 찾아봤지만 조리법이 없으므로 공백으로 놔두겠음.

Name                                        Ingredients  \
1940  Vodka & Tonic (VAT)  1 2⁄3 oz Ketel One Vodka, 4 oz Thomas Henry To...   
5227      Green Tea Punch  1 2⁄3 oz Drambuie liqueur, 2 oz Green tea (col...   

     Making                                               Link          Glass  \
1940    NaN  https://www.diffordsguide.com/cocktails/recipe...  Collins glass   
5227    NaN  https://www.diffordsguide.com/cocktails/recipe...            Cup   

                      Garnish Type  \
1940               Lime wedge  NaN   
5227  Lemon & cucumber slices  NaN   

                                                 Flavor  
1940  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...  
5227                                                NaN

In [369]:
difcopy["Making"][6]

'Cut zest from your lime for garnish prior to squeezing for juice. SHAKE all ingredients with ice and fine strain into chilled glass.\n\nFor an even better Daiquiri use 2 x 5ml measuring spoons of powdered sugar (finely ground caster sugar) in place of 10ml sugar syrup. Stir sugar with other ingredients in base of shaker prior to adding ice and shaking to ensure fully dissolved.'

In [370]:
for i in range(len(difcopy)):
    if i!=1940 and i!=5227:
        making = difcopy["Making"][i].replace("\n", " ")
        making = making.replace("  ", " ")
        difcopy["Making"][i] = making

In [371]:
difcopy["Making"][6]

'Cut zest from your lime for garnish prior to squeezing for juice. SHAKE all ingredients with ice and fine strain into chilled glass. For an even better Daiquiri use 2 x 5ml measuring spoons of powdered sugar (finely ground caster sugar) in place of 10ml sugar syrup. Stir sugar with other ingredients in base of shaker prior to adding ice and shaking to ensure fully dissolved.'

In [372]:
for i in range(len(difcopy)):
    if str(difcopy["Making"][i])=='nan':
        difcopy["Making"][i]=None

### 이름이 다른데 재료가 같은 칵테일의 다른 속성들을 보고 중복인지 아닌지 판단하기

In [373]:
difcopy_dup = difcopy.duplicated(["Ingredients"], keep=False)
len(difcopy_dup[difcopy_dup==True])

10

In [374]:
difcopy_dup_ind = []
for i in range(len(difcopy)):
    if difcopy_dup[i]==True:
        difcopy_dup_ind.append(i)
print(difcopy_dup_ind)

[484, 525, 1566, 2625, 4114, 4193, 4276, 4277, 4922, 4973]


In [375]:
for i in difcopy_dup_ind:
    print(i)
    print(difcopy["Name"][i])
    print(difcopy["Making"][i])
    print(difcopy["Ingredients"][i])
    print(difcopy["Glass"][i])
    print(difcopy["Garnish"][i])
    print("ㅡㅡㅡㅡㅡ")

484
Bohemian Martini
STIR all ingredients with ice and strain into chilled glass.
2 1⁄2 oz Rutte Dry Gin, 1⁄2 oz Martini Extra Dry vermouth
Martini glass
Caperberry
ㅡㅡㅡㅡㅡ
525
Buckeye
STIR all ingredients with ice and strain into chilled glass.
2 1⁄2 oz Rutte Dry Gin, 1⁄2 oz Martini Extra Dry vermouth
Martini glass
Black olive
ㅡㅡㅡㅡㅡ
1566
Roosevelt Martini
STIR all ingredients with ice and strain into chilled glass.
2 1⁄2 oz Rutte Dry Gin, 1⁄2 oz Martini Extra Dry vermouth
Martini glass
Two olives on stick
ㅡㅡㅡㅡㅡ
2625
Eggnog (hot)
POUR ingredients into heatproof glass and STIR thoroughly. HEAT in microwave oven for a minute (adjust time as appropriate to your oven) and STIR again. Alternatively, mix and warm in pan over heat - do not boil.
2 1⁄2 oz Rémy Martin 1738 Cognac, "1⁄2 oz Sugar syrup 'rich' (2 sugar to 1 water, 65.0°brix)", 1 fresh Egg (white & yolk), 1⁄2 oz Single cream / half-and-half, 2 oz Milk (whole milk/full 3-4% fat)
Toddy glass
Dust with grated nutmeg
ㅡㅡㅡㅡㅡ
4114
Bartender

(4114, 4193, 4922), (4276, 4277)는 이름만 다르고 레시피가 같은 칵테일들이다. 하지만 아직 부여하지 않은 속성들이 있고 다른 사이트의 데이터에도 같은 칵테일이 있을지도 모르므로 데이터를 취합한 후에 다시 확인해서 처리할 것.

### Glass 형식 맞추기

In [376]:
for i in range(len(difcopy)):
    if str(difcopy["Glass"][i])=='nan':
        difcopy["Glass"][i]=None

In [377]:
difcopy["Glass"].unique()

array(['Martini glass', 'Coupe glass', 'Old-fashioned glass',
       'Collins glass', 'Highball (max 10oz/300ml)',
       'Martini (small) glass', 'Nick & Nora glass', 'Julep tin',
       'Rocks glass', 'Flute glass', 'Sour or Martini/Coupette glass',
       'Double old-fashioned', 'Toddy glass', 'Shot glass', 'Poco grande',
       'Cup', 'Sling glass', 'Wine glass', 'Tiki mug or collins',
       'Copa glass', 'Snifter glass', 'Martini (large 10oz) glass',
       'Hurricane glass', 'Cantaritos clay pot or Collins glass',
       'Large wine glass', 'Copita sherry glass',
       'Cartridge mug or Collins glass', 'Goblet glass',
       'Shot and Beer glass', 'Copper mug or Collins glass',
       'Pineapple shell (frozen) glass', 'Belgium beer glass',
       'Rum barrel mug or pint glass', 'Fizz or Highball (8oz to 10oz)',
       'Absinthe glass or old-fashioned glass', 'Jam jar glass',
       'Daddy Rack Tennessee Whiskey', 'Coconut shell or Collins glass',
       'BarSol Mosto Verde Ital

In [378]:
for i in range(len(difcopy)):
    if difcopy["Glass"][i].split(" ")[-1]=='glass':
        difcopy["Glass"][i] = difcopy["Glass"][i].replace("glass", "Glass")

In [379]:
difcopy["Glass"].unique()

array(['Martini Glass', 'Coupe Glass', 'Old-fashioned Glass',
       'Collins Glass', 'Highball (max 10oz/300ml)',
       'Martini (small) Glass', 'Nick & Nora Glass', 'Julep tin',
       'Rocks Glass', 'Flute Glass', 'Sour or Martini/Coupette Glass',
       'Double old-fashioned', 'Toddy Glass', 'Shot Glass', 'Poco grande',
       'Cup', 'Sling Glass', 'Wine Glass', 'Tiki mug or collins',
       'Copa Glass', 'Snifter Glass', 'Martini (large 10oz) Glass',
       'Hurricane Glass', 'Cantaritos clay pot or Collins Glass',
       'Large wine Glass', 'Copita sherry Glass',
       'Cartridge mug or Collins Glass', 'Goblet Glass',
       'Shot and Beer Glass', 'Copper mug or Collins Glass',
       'Pineapple shell (frozen) Glass', 'Belgium beer Glass',
       'Rum barrel mug or pint Glass', 'Fizz or Highball (8oz to 10oz)',
       'Absinthe Glass or old-fashioned Glass', 'Jam jar Glass',
       'Daddy Rack Tennessee Whiskey', 'Coconut shell or Collins Glass',
       'BarSol Mosto Verde Ital

In [380]:
for i in range(len(difcopy)):
    if difcopy["Glass"][i]=='Old-fashioned Glass' or difcopy["Glass"][i]=='Double old-fashioned':
        difcopy["Glass"][i] = 'Old Fashioned Glass'
    if difcopy["Glass"][i]=='Highball (max 10oz/300ml)':
        difcopy["Glass"][i] = 'Highball Glass'
    if difcopy["Glass"][i]=='Martini (small) Glass' or difcopy["Glass"][i]=='Martini (large 10oz) Glass':
        difcopy["Glass"][i] = 'Martini Glass'
    if difcopy["Glass"][i]=='Julep tin':
        difcopy["Glass"][i] = 'Julep Tin'
    if difcopy["Glass"][i]=='Sour or Martini/Coupette Glass':
        difcopy["Glass"][i] = 'Sour or Martini or Coupette Glass'
    if difcopy["Glass"][i]=='Poco grande':
        difcopy["Glass"][i] = 'Poco Grande Glass'
    if difcopy["Glass"][i]=='Tiki mug or collins':
        difcopy["Glass"][i] = 'Tiki Mug or Collins Glass'
    if difcopy["Glass"][i]=='Cantaritos clay pot or Collins Glass':
        difcopy["Glass"][i] = 'Cantaritos Clay Pot or Collins Glass'
    if difcopy["Glass"][i]=='Large wine Glass':
        difcopy["Glass"][i] = 'Wine Glass'
    if difcopy["Glass"][i]=='Copita sherry Glass':
        difcopy["Glass"][i] = 'Copita Sherry Glass'
    if difcopy["Glass"][i]=='Cartridge mug or Collins Glass':
        difcopy["Glass"][i] = 'Cartridge Mug or Collins Glass'
    if difcopy["Glass"][i]=='Shot and Beer Glass':
        difcopy["Glass"][i] = 'Shot or Beer Glass'
    if difcopy["Glass"][i]=='Copper mug or Collins Glass':
        difcopy["Glass"][i] = 'Copper Mug or Collins Glass'
    if difcopy["Glass"][i]=='Pineapple shell (frozen) Glass':
        difcopy["Glass"][i] = 'Pineapple Shell or Glass'
    if difcopy["Glass"][i]=='Belgium beer Glass':
        difcopy["Glass"][i] = 'Belgium Beer Glass'
    if difcopy["Glass"][i]=='Rum barrel mug or pint Glass':
        difcopy["Glass"][i] = 'Rum Barrel Mug or Pint Glass'
    if difcopy["Glass"][i]=='Fizz or Highball (8oz to 10oz)':
        difcopy["Glass"][i] = 'Fizz or Highball Glass'
    if difcopy["Glass"][i]=='Absinthe Glass or old-fashioned Glass':
        difcopy["Glass"][i] = 'Absinthe or Old Fashioned Glass'
    if difcopy["Glass"][i]=='Jam jar Glass':
        difcopy["Glass"][i] = 'Jam Jar Glass'
    if difcopy["Glass"][i]=='Daddy Rack Tennessee Whiskey':
        difcopy["Glass"][i] = None
    if difcopy["Glass"][i]=='Coconut shell or Collins Glass':
        difcopy["Glass"][i] = 'Coconut Shell or Collins Glass'
    if difcopy["Glass"][i]=='BarSol Mosto Verde Italia Pisco':
        difcopy["Glass"][i] = None
    if difcopy["Glass"][i]=='Medical cup or Old-fashioned Glass':
        difcopy["Glass"][i] = 'Medical Cup or Old Fashioned Glass'
    if difcopy["Glass"][i]=='Collins or Pineapple shell Glass':
        difcopy["Glass"][i] = 'Pineapple Shell or Collins Glass'
    if difcopy["Glass"][i]=='Boston & shot Glass':
        difcopy["Glass"][i] = 'Boston or Shot Glass'
    if difcopy["Glass"][i]=='Espresso cup':
        difcopy["Glass"][i] = 'Espresso Cup'

In [381]:
difcopy["Glass"].unique()

array(['Martini Glass', 'Coupe Glass', 'Old Fashioned Glass',
       'Collins Glass', 'Highball Glass', 'Nick & Nora Glass',
       'Julep Tin', 'Rocks Glass', 'Flute Glass',
       'Sour or Martini or Coupette Glass', 'Toddy Glass', 'Shot Glass',
       'Poco Grande Glass', 'Cup', 'Sling Glass', 'Wine Glass',
       'Tiki Mug or Collins Glass', 'Copa Glass', 'Snifter Glass',
       'Hurricane Glass', 'Cantaritos Clay Pot or Collins Glass',
       'Copita Sherry Glass', 'Cartridge Mug or Collins Glass',
       'Goblet Glass', 'Shot or Beer Glass',
       'Copper Mug or Collins Glass', 'Pineapple Shell or Glass',
       'Belgium Beer Glass', 'Rum Barrel Mug or Pint Glass',
       'Fizz or Highball Glass', 'Absinthe or Old Fashioned Glass',
       'Jam Jar Glass', None, 'Coconut Shell or Collins Glass',
       'Medical Cup or Old Fashioned Glass',
       'Pineapple Shell or Collins Glass', 'Pint Glass', 'Boston Glass',
       'Boston or Shot Glass', 'Margarita Glass', 'Tin', 'Espresso Cu

### Garnish 형식 맞추기

In [382]:
for i in range(len(difcopy)):
    if str(difcopy["Glass"][i])=='nan':
        difcopy["Glass"][i]=None

In [383]:
for i in range(len(difcopy)):
    if str(difcopy["Garnish"][i])!='None':
        if "\n" in difcopy["Garnish"][i]:
            print(i)
# 출력된 세 개의 레시피를 자세히 살펴보니
# Garnish 속성에 Making이 들어가버린 자료였음
# 공통점: Making에 'Read about cocktail measures and measuring'이 써있음, 원래 사이트에서 Garnish란이 아예 없음
# 처리해주기

2207
3206
5257


In [384]:
for i in range(len(difcopy)):
    if str(difcopy["Making"][i])!='None':
        if 'Read about cocktail measures and measuring' in difcopy["Making"][i]:
            print(i)

822
2207
3206
3384
4319
5257


In [385]:
difcopy["Making"][822]=difcopy["Garnish"][822]
difcopy["Garnish"][822]=None
difcopy["Making"][2207]=difcopy["Garnish"][2207]
difcopy["Garnish"][2207]=None
difcopy["Making"][3206]=difcopy["Garnish"][3206]
difcopy["Garnish"][3206]=None
difcopy["Making"][3384]=difcopy["Garnish"][3384]
difcopy["Garnish"][3384]=None
difcopy["Making"][4319]=difcopy["Garnish"][4319]
difcopy["Garnish"][4319]=None
difcopy["Making"][5257]=difcopy["Garnish"][5257]
difcopy["Garnish"][5257]=None

Making 형식 다시 맞추기

In [386]:
for i in range(len(difcopy)):
    if i!=1940 and i!=5227:
        making = difcopy["Making"][i].replace("\n", " ")
        making = making.replace("  ", " ")
        difcopy["Making"][i] = making

Garnish 형식은 맞출 필요가 없어졌으니 넘어가기

### Base 속성 부여하기(이 사이트는 Base 데이터를 따로 스크랩 할 수 있었음)

In [387]:
dif_beer = pd.read_csv("//Difford's_beer.csv")
beer_list = list(dif_beer["Name"])
dif_brandy = pd.read_csv("//Difford's_brandy.csv")
brandy_list = list(dif_brandy["Name"])
dif_gin = pd.read_csv("//Difford's_gin.csv")
gin_list = list(dif_gin["Name"])
dif_redwine = pd.read_csv("//Difford's_redwine.csv")
redwine_list = list(dif_redwine["Name"])
dif_rum = pd.read_csv("//Difford's_rum.csv")
rum_list = list(dif_rum["Name"])
dif_sherry = pd.read_csv("//Difford's_sherry.csv")
sherry_list = list(dif_sherry["Name"])
dif_tequila = pd.read_csv("//Difford's_tequila.csv")
tequila_list = list(dif_tequila["Name"])
dif_vodka = pd.read_csv("//Difford's_vodka.csv")
vodka_list = list(dif_vodka["Name"])
dif_whisky = pd.read_csv("//Difford's_whisky.csv")
whisky_list = list(dif_whisky["Name"])
dif_whitewine = pd.read_csv("//Difford's_whitewine.csv")
whitewine_list = list(dif_whitewine["Name"])

In [388]:
print(beer_list)

['Beer on-the-rocks', 'De Beauvoir', 'Anything Else', "Bitches' Brew cocktail", 'Boilermaker', 'Czechmate', 'Dunkel & Stürmig', 'Flip the Script', 'Hwachae', 'Jack Pot', 'Kopstoot', 'Mantini', 'Reality Check', 'Revolutionary Flip', 'Swedish Ale Punch', 'The Angel Wore Red', 'Weissen Sour', 'Am I Blue?', "Beggar's Banquet", 'Black Velvet', 'Drop By Drop', 'El Camello Loco', 'El Tindero', "Gilbey Vintner's Punch", 'Huckle my buff', 'IPAlicus', 'Irish Shandy', 'Michelada', 'Pale Charlie', 'Reborn Padre Blanco', 'Seventh Son', 'Shaky Pete’s Ginger Brew', 'Smoky Mocha Spritz', 'South Ireland Sour', 'Stout Fellow', 'Black & Tan', 'Black & Velvet', 'Peroni Negroni', 'Sandygaff', 'Blackbeard', 'Doctor Pepper', 'Shandygaff', 'Snakebite', 'Steel Bottom', 'The Purl', 'Black Japan', 'Depth Charge', 'Flaming Dr Pepper', "Clever Jasper's Tea", 'Hug From Mary']


In [389]:
type(difcopy["Name"][1])

str

In [390]:
base_list = [None]*5339
for i in range(len(difcopy)):
    bases = []
    if difcopy["Name"][i] in beer_list:
        bases.append("Beer")
    if difcopy["Name"][i] in brandy_list:
        bases.append("Brandy")
    if difcopy["Name"][i] in gin_list:
        bases.append("Gin")
    if difcopy["Name"][i] in redwine_list:
        bases.append("Red Wine")
    if difcopy["Name"][i] in rum_list:
        bases.append("Rum")
    if difcopy["Name"][i] in sherry_list:
        bases.append("Sherry")
    if difcopy["Name"][i] in tequila_list:
        bases.append("Tequila")
    if difcopy["Name"][i] in vodka_list:
        bases.append("Vodka")
    if difcopy["Name"][i] in whisky_list:
        bases.append("Whisky")
    if difcopy["Name"][i] in whitewine_list:
        bases.append("White Wine")
    if "mezcal" in difcopy["Ingredients"][i].lower():
        bases.append("Mezcal")
    if "cognac" in difcopy["Ingredients"][i].lower():
        bases.append("Cognac")
    if " ale," in difcopy["Ingredients"][i].lower() or " ale." in difcopy["Ingredients"][i].lower() or " ale " in difcopy["Ingredients"][i].lower():
        bases.append("Ale")
    if "champagne" in difcopy["Ingredients"][i].lower():
        bases.append("Champagne")
    if bases==[]:
        pass
    else:
        base_list[i] = ", ".join(bases)

In [391]:
base_list.count(None)

1028

In [392]:
difcopy["Base"]=base_list

In [393]:
# 확인
difcopy.head()

Name  \
0                                April Fool   
1                        Brooklyn (perfect)   
2                     Californian Margarita   
3  Calvados Fruit Cup (Difford's Cup No.13)   
4                  Clover Club (House-Made)   

                                         Ingredients  \
0                                 3 oz Chilled water   
1  1 1⁄2 oz Straight rye whiskey (100 proof / 50%...   
2  1 1⁄2 oz Patrón Silver blanco tequila, 3⁄4 oz ...   
3  5⁄6 oz Avallen Calvados, 5⁄6 oz Orange Curaçao...   
4  1 1⁄2 oz Rutte Dry Gin, 1 1⁄2 spoon Raspberry ...   

                                              Making  \
0       STIR with ice and strain into chilled glass.   
1  STIR all ingredients with ice and strain into ...   
2  SHAKE all ingredients with ice and strain into...   
3  POUR all ingredients into glass. Half fill gla...   
4  STIR jam with gin in base of shaker until diss...   

                                                Link                Glass  \
0  https://www.diffordsguide.com/cocktails/recipe...        Martini Glass   
1  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   
2  https://www.diffordsguide.com/cocktails/recipe...  Old Fashioned Glass   
3  https://www.diffordsguide.com/cocktails/recipe...        Collins Glass   
4  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   

                                             Garnish          Type  \
0               Lemon zest twist (discarded) & olive  Nonalcoholic   
1                          Luxardo maraschino cherry           NaN   
2                   Salt rim (optional) & lime wedge           NaN   
3  Lemon, orange & strawberry slices, mint sprig ...           NaN   
4  Lemon zest twist (expressed and discarded) & s...           NaN   

                                              Flavor         Base  
0                                                NaN         None  
1  ['Gentle 0 - Boozy 10: 8', 'Sweet 0 - Dry/Sour...       Whisky  
2  ['Gentle 0 - Boozy 10: 7', 'Sweet 0 - Dry/Sour...      Tequila  
3  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...  Brandy, Ale  
4  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...          Gin

### Method 추출하기

In [394]:
method_list = [None]*5339
for i in range(len(difcopy["Making"])):
    methods = []
    if str(difcopy["Making"][i])!='None':
        if "shake" in difcopy["Making"][i].lower() or "shkae" in difcopy["Making"][i].lower():
            methods.append("Shake")
        if "stir" in difcopy["Making"][i].lower():
            if "build" in difcopy["Making"][i].lower():
                methods.append("Build")
            elif "combine" in difcopy["Making"][i].lower():
                methods.append("Build")
            else:
                methods.append("Stir")
        elif "build" in difcopy["Making"][i].lower():
            methods.append("Build")
        elif "combine" in difcopy["Making"][i].lower():
            methods.append("Build")
        elif "mix over" in difcopy["Making"][i].lower():
            methods.append("Stir")
        elif "mix in" in difcopy["Making"][i].lower():
            methods.append("Stir")
        elif "mix with" in difcopy["Making"][i].lower():
            methods.append("Stir")
        elif "mix well" in difcopy["Making"][i].lower():
            methods.append("Stir")
        elif "mix together" in difcopy["Making"][i].lower():
            methods.append("Stir")
        elif "mix all" in difcopy["Making"][i].lower():
            methods.append("Stir")
        elif "mixing glass" in difcopy["Making"][i].lower():
            methods.append("Stir")
        if "float" in difcopy["Making"][i].lower() or "layer" in difcopy["Making"][i].lower():
            methods.append("Float")
        if "blend" in difcopy["Making"][i].lower():
            methods.append("Blend")
        if "strain" in difcopy["Making"][i].lower():
            methods.append("Strain")
    methods = list(set(methods))
    if methods==[]:
        pass
    else:
        method_list[i]=", ".join(methods)

Method 속성 부여하고 확인하기

In [395]:
difcopy["Method"]=method_list
difcopy.head()

Name  \
0                                April Fool   
1                        Brooklyn (perfect)   
2                     Californian Margarita   
3  Calvados Fruit Cup (Difford's Cup No.13)   
4                  Clover Club (House-Made)   

                                         Ingredients  \
0                                 3 oz Chilled water   
1  1 1⁄2 oz Straight rye whiskey (100 proof / 50%...   
2  1 1⁄2 oz Patrón Silver blanco tequila, 3⁄4 oz ...   
3  5⁄6 oz Avallen Calvados, 5⁄6 oz Orange Curaçao...   
4  1 1⁄2 oz Rutte Dry Gin, 1 1⁄2 spoon Raspberry ...   

                                              Making  \
0       STIR with ice and strain into chilled glass.   
1  STIR all ingredients with ice and strain into ...   
2  SHAKE all ingredients with ice and strain into...   
3  POUR all ingredients into glass. Half fill gla...   
4  STIR jam with gin in base of shaker until diss...   

                                                Link                Glass  \
0  https://www.diffordsguide.com/cocktails/recipe...        Martini Glass   
1  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   
2  https://www.diffordsguide.com/cocktails/recipe...  Old Fashioned Glass   
3  https://www.diffordsguide.com/cocktails/recipe...        Collins Glass   
4  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   

                                             Garnish          Type  \
0               Lemon zest twist (discarded) & olive  Nonalcoholic   
1                          Luxardo maraschino cherry           NaN   
2                   Salt rim (optional) & lime wedge           NaN   
3  Lemon, orange & strawberry slices, mint sprig ...           NaN   
4  Lemon zest twist (expressed and discarded) & s...           NaN   

                                              Flavor         Base  \
0                                                NaN         None   
1  ['Gentle 0 - Boozy 10: 8', 'Sweet 0 - Dry/Sour...       Whisky   
2  ['Gentle 0 - Boozy 10: 7', 'Sweet 0 - Dry/Sour...      Tequila   
3  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...  Brandy, Ale   
4  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...          Gin   

                Method  
0         Stir, Strain  
1         Stir, Strain  
2        Shake, Strain  
3                 None  
4  Shake, Strain, Stir

### Type 추출하기

In [396]:
difcopy["Glass"].unique()

array(['Martini Glass', 'Coupe Glass', 'Old Fashioned Glass',
       'Collins Glass', 'Highball Glass', 'Nick & Nora Glass',
       'Julep Tin', 'Rocks Glass', 'Flute Glass',
       'Sour or Martini or Coupette Glass', 'Toddy Glass', 'Shot Glass',
       'Poco Grande Glass', 'Cup', 'Sling Glass', 'Wine Glass',
       'Tiki Mug or Collins Glass', 'Copa Glass', 'Snifter Glass',
       'Hurricane Glass', 'Cantaritos Clay Pot or Collins Glass',
       'Copita Sherry Glass', 'Cartridge Mug or Collins Glass',
       'Goblet Glass', 'Shot or Beer Glass',
       'Copper Mug or Collins Glass', 'Pineapple Shell or Glass',
       'Belgium Beer Glass', 'Rum Barrel Mug or Pint Glass',
       'Fizz or Highball Glass', 'Absinthe or Old Fashioned Glass',
       'Jam Jar Glass', None, 'Coconut Shell or Collins Glass',
       'Medical Cup or Old Fashioned Glass',
       'Pineapple Shell or Collins Glass', 'Pint Glass', 'Boston Glass',
       'Boston or Shot Glass', 'Margarita Glass', 'Tin', 'Espresso Cu

추출할 속성: short, long, shooter<br>
바꿔줄 속성: "Nonalcoholic" to "Non-Alcoholic"

In [397]:
type_list = [None]*5339
shooter_glasses = ['Shot Glass']
short_glasses = ['Martini Glass', 'Coupe Glass', 'Sour or Martini or Coupette Glass']
long_glasses = ['Collins Glass', 'Highball Glass', 'Sling Glass', 'Tiki Mug or Collins Glass', 'Hurricane Glass', 'Cantaritos Clay Pot or Collins Glass', 'Copper Mug or Collins Glass', 'Fizz or Highball Glass']
for i in range(len(difcopy)):
    count = 0
    types = []
    if difcopy["Type"][i]=="Nonalcoholic":
        types.append("Non-Alcoholic")
    if difcopy["Glass"][i] in shooter_glasses:
        types.append("Shooter")
        count += 1
    if difcopy["Glass"][i] in short_glasses:
        types.append("Short")
        count += 1
    if difcopy["Glass"][i] in long_glasses:
        types.append("Long")
        count += 1
    if count>=2:
        print('here!', i)
        if "Non-Alcoholic" in types:
            types = ["Non-Alcoholic"]
        else:
            types = []
    if types==[]:
        pass
    else:
        type_list[i]=", ".join(types)

Type 속성 부여하고 확인하기

In [398]:
difcopy["Type"]=type_list
difcopy.head()

Name  \
0                                April Fool   
1                        Brooklyn (perfect)   
2                     Californian Margarita   
3  Calvados Fruit Cup (Difford's Cup No.13)   
4                  Clover Club (House-Made)   

                                         Ingredients  \
0                                 3 oz Chilled water   
1  1 1⁄2 oz Straight rye whiskey (100 proof / 50%...   
2  1 1⁄2 oz Patrón Silver blanco tequila, 3⁄4 oz ...   
3  5⁄6 oz Avallen Calvados, 5⁄6 oz Orange Curaçao...   
4  1 1⁄2 oz Rutte Dry Gin, 1 1⁄2 spoon Raspberry ...   

                                              Making  \
0       STIR with ice and strain into chilled glass.   
1  STIR all ingredients with ice and strain into ...   
2  SHAKE all ingredients with ice and strain into...   
3  POUR all ingredients into glass. Half fill gla...   
4  STIR jam with gin in base of shaker until diss...   

                                                Link                Glass  \
0  https://www.diffordsguide.com/cocktails/recipe...        Martini Glass   
1  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   
2  https://www.diffordsguide.com/cocktails/recipe...  Old Fashioned Glass   
3  https://www.diffordsguide.com/cocktails/recipe...        Collins Glass   
4  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   

                                             Garnish                  Type  \
0               Lemon zest twist (discarded) & olive  Non-Alcoholic, Short   
1                          Luxardo maraschino cherry                 Short   
2                   Salt rim (optional) & lime wedge                  None   
3  Lemon, orange & strawberry slices, mint sprig ...                  Long   
4  Lemon zest twist (expressed and discarded) & s...                 Short   

                                              Flavor         Base  \
0                                                NaN         None   
1  ['Gentle 0 - Boozy 10: 8', 'Sweet 0 - Dry/Sour...       Whisky   
2  ['Gentle 0 - Boozy 10: 7', 'Sweet 0 - Dry/Sour...      Tequila   
3  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...  Brandy, Ale   
4  ['Gentle 0 - Boozy 10: 6', 'Sweet 0 - Dry/Sour...          Gin   

                Method  
0         Stir, Strain  
1         Stir, Strain  
2        Shake, Strain  
3                 None  
4  Shake, Strain, Stir

### Flavor 형식 맞추기

In [399]:
for i in range(len(difcopy)):
    if str(difcopy["Flavor"][i])=='nan':
        difcopy["Flavor"][i]=None
    else:
        flavors = difcopy["Flavor"][i][1:-1].split(", ")
        for j in range(len(flavors)):
            if flavors[j][0]=="'":
                flavors[j] = flavors[j][1:]
            if flavors[j][-1]=="'":
                flavors[j] = flavors[j][:-1]
        flavor = ", ".join(flavors)
        difcopy["Flavor"][i] = flavor

In [400]:
# 확인
difcopy.head()

Name  \
0                                April Fool   
1                        Brooklyn (perfect)   
2                     Californian Margarita   
3  Calvados Fruit Cup (Difford's Cup No.13)   
4                  Clover Club (House-Made)   

                                         Ingredients  \
0                                 3 oz Chilled water   
1  1 1⁄2 oz Straight rye whiskey (100 proof / 50%...   
2  1 1⁄2 oz Patrón Silver blanco tequila, 3⁄4 oz ...   
3  5⁄6 oz Avallen Calvados, 5⁄6 oz Orange Curaçao...   
4  1 1⁄2 oz Rutte Dry Gin, 1 1⁄2 spoon Raspberry ...   

                                              Making  \
0       STIR with ice and strain into chilled glass.   
1  STIR all ingredients with ice and strain into ...   
2  SHAKE all ingredients with ice and strain into...   
3  POUR all ingredients into glass. Half fill gla...   
4  STIR jam with gin in base of shaker until diss...   

                                                Link                Glass  \
0  https://www.diffordsguide.com/cocktails/recipe...        Martini Glass   
1  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   
2  https://www.diffordsguide.com/cocktails/recipe...  Old Fashioned Glass   
3  https://www.diffordsguide.com/cocktails/recipe...        Collins Glass   
4  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   

                                             Garnish                  Type  \
0               Lemon zest twist (discarded) & olive  Non-Alcoholic, Short   
1                          Luxardo maraschino cherry                 Short   
2                   Salt rim (optional) & lime wedge                  None   
3  Lemon, orange & strawberry slices, mint sprig ...                  Long   
4  Lemon zest twist (expressed and discarded) & s...                 Short   

                                             Flavor         Base  \
0                                              None         None   
1  Gentle 0 - Boozy 10: 8, Sweet 0 - Dry/Sour 10: 6       Whisky   
2  Gentle 0 - Boozy 10: 7, Sweet 0 - Dry/Sour 10: 7      Tequila   
3  Gentle 0 - Boozy 10: 6, Sweet 0 - Dry/Sour 10: 6  Brandy, Ale   
4  Gentle 0 - Boozy 10: 6, Sweet 0 - Dry/Sour 10: 8          Gin   

                Method  
0         Stir, Strain  
1         Stir, Strain  
2        Shake, Strain  
3                 None  
4  Shake, Strain, Stir

카테고리 순서 형식에 맞게 정렬하고 확인하기

In [401]:
difcopy = difcopy[["Name", "Ingredients", "Making", "Link", "Glass", "Garnish", "Base", "Method", "Type", "Flavor"]]
difcopy.head()

Name  \
0                                April Fool   
1                        Brooklyn (perfect)   
2                     Californian Margarita   
3  Calvados Fruit Cup (Difford's Cup No.13)   
4                  Clover Club (House-Made)   

                                         Ingredients  \
0                                 3 oz Chilled water   
1  1 1⁄2 oz Straight rye whiskey (100 proof / 50%...   
2  1 1⁄2 oz Patrón Silver blanco tequila, 3⁄4 oz ...   
3  5⁄6 oz Avallen Calvados, 5⁄6 oz Orange Curaçao...   
4  1 1⁄2 oz Rutte Dry Gin, 1 1⁄2 spoon Raspberry ...   

                                              Making  \
0       STIR with ice and strain into chilled glass.   
1  STIR all ingredients with ice and strain into ...   
2  SHAKE all ingredients with ice and strain into...   
3  POUR all ingredients into glass. Half fill gla...   
4  STIR jam with gin in base of shaker until diss...   

                                                Link                Glass  \
0  https://www.diffordsguide.com/cocktails/recipe...        Martini Glass   
1  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   
2  https://www.diffordsguide.com/cocktails/recipe...  Old Fashioned Glass   
3  https://www.diffordsguide.com/cocktails/recipe...        Collins Glass   
4  https://www.diffordsguide.com/cocktails/recipe...          Coupe Glass   

                                             Garnish         Base  \
0               Lemon zest twist (discarded) & olive         None   
1                          Luxardo maraschino cherry       Whisky   
2                   Salt rim (optional) & lime wedge      Tequila   
3  Lemon, orange & strawberry slices, mint sprig ...  Brandy, Ale   
4  Lemon zest twist (expressed and discarded) & s...          Gin   

                Method                  Type  \
0         Stir, Strain  Non-Alcoholic, Short   
1         Stir, Strain                 Short   
2        Shake, Strain                  None   
3                 None                  Long   
4  Shake, Strain, Stir                 Short   

                                             Flavor  
0                                              None  
1  Gentle 0 - Boozy 10: 8, Sweet 0 - Dry/Sour 10: 6  
2  Gentle 0 - Boozy 10: 7, Sweet 0 - Dry/Sour 10: 7  
3  Gentle 0 - Boozy 10: 6, Sweet 0 - Dry/Sour 10: 6  
4  Gentle 0 - Boozy 10: 6, Sweet 0 - Dry/Sour 10: 8

데이터 저장하기

In [402]:
difcopy.to_csv("//Diffords_ready.csv", encoding="utf-8-sig", index=False)

Difford's Guide 데이터 전처리 완료